File: src/test/java/com/github/nsc/de/shake/CompilerError.java
package com.github.nsc.de.shake;

import com.github.nsc.de.shake.lexer.characterinputstream.charactersource.CharacterSource;
import com.github.nsc.de.shake.lexer.characterinputstream.position.Position;
import com.github.nsc.de.shake.lexer.characterinputstream.position.PositionMap;
import com.github.nsc.de.shake.util.Formatting;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.*;

public class CompilerError {

    @Test
    public void testCompilerError() {
        CharacterSource source = CharacterSource.from(genLengthString(30), "<source>");
        PositionMap map = new PositionMap(source, new int[] {});

        com.github.nsc.de.shake.util.CompilerError error =
                new com.github.nsc.de.shake.util.CompilerError(
                        "message", "TestingError", "Some details",
                        map.resolve(10), map.resolve(10));

        assertEquals("<source>:1:11", error.getMarker().getSource());
        assertEquals("1  012345678901234567890123456789", error.getMarker().getPreview());
        assertEquals("             ^", error.getMarker().getMarker());

        assertEquals("1  0123456789" + Formatting.INVERT + Formatting.FGColor.RED + "0" +
                Formatting.RESET + "1234567890123456789", error.getMarker().getColorPreview());

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);

    }

    @Test
    public void testCompilerErrorOverflowAfter() {
        CharacterSource source = CharacterSource.from(genLengthString(60), "<source>");
        PositionMap map = new PositionMap(source, new int[] {});

        com.github.nsc.de.shake.util.CompilerError error =
                new com.github.nsc.de.shake.util.CompilerError(
                        "message", "TestingError", "Some details",
                        new Position(map, 10, 11, 1),
                        new Position(map, 10, 11, 1));

        // System.out.println(error.toString());

        assertEquals("<source>:1:11", error.getMarker().getSource());
        assertEquals("1  0123456789012345678901234567890...+28", error.getMarker().getPreview());
        assertEquals("             ^", error.getMarker().getMarker());

        assertEquals("1  0123456789" + Formatting.INVERT + Formatting.FGColor.RED + "0" +
                Formatting.RESET + "12345678901234567890...+28", error.getMarker().getColorPreview());

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);

    }

    @Test
    public void testCompilerErrorOverflowBefore() {
        CharacterSource source = CharacterSource.from(genLengthString(60), "<source>");
        PositionMap map = new PositionMap(source, new int[] {});

        com.github.nsc.de.shake.util.CompilerError error =
                new com.github.nsc.de.shake.util.CompilerError(
                        "message", "TestingError", "Some details",
                        map.resolve(39), map.resolve(39));

        // System.out.println(error.toString());

        assertEquals("<source>:1:40", error.getMarker().getSource());
        assertEquals("1  +18...890123456789012345678901234567890123456789", error.getMarker().getPreview());
        assertEquals("                              ^", error.getMarker().getMarker());

        assertEquals("1  +18...890123456789012345678" + Formatting.INVERT + Formatting.FGColor.RED + "9" +
                Formatting.RESET + "01234567890123456789", error.getMarker().getColorPreview());

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);

    }

    @Test
    public void testCompilerErrorOverflow() {
        CharacterSource source = CharacterSource.from(genLengthString(100), "<source>");
        PositionMap map = new PositionMap(source, new int[] {});

        com.github.nsc.de.shake.util.CompilerError error =
                new com.github.nsc.de.shake.util.CompilerError(
                        "message", "TestingError", "Some details",
                        map.resolve(49), map.resolve(49));

        // System.out.println(error.toString());

        assertEquals("<source>:1:50", error.getMarker().getSource());
        assertEquals("1  +28...890123456789012345678901234567890123456789...+29", error.getMarker().getPreview());
        assertEquals("                              ^", error.getMarker().getMarker());

        assertEquals("1  +28...890123456789012345678" + Formatting.INVERT + Formatting.FGColor.RED + "9" +
                Formatting.RESET + "01234567890123456789...+29", error.getMarker().getColorPreview());

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);

    }

    @Test
    public void testCompilerErrorLongMark() {
        CharacterSource source = CharacterSource.from(genLengthString(40), "<source>");
        PositionMap map = new PositionMap(source, new int[] {});

        com.github.nsc.de.shake.util.CompilerError error =
                new com.github.nsc.de.shake.util.CompilerError(
                        "message", "TestingError", "Some details",
                        map.resolve(9),
                        map.resolve(14));

        // System.out.println(error.toString());

        assertEquals("<source>:1:10", error.getMarker().getSource());
        assertEquals("1  012345678901234567890123456789012345678", error.getMarker().getPreview());
        assertEquals("            ^^^^^^", error.getMarker().getMarker());

        assertEquals("1  012345678" + Formatting.INVERT + Formatting.FGColor.RED + "901234" +
                Formatting.RESET + "567890123456789012345678", error.getMarker().getColorPreview());

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);

    }

    @Test
    public void testCompilerErrorMultiLine() {
        String str = '\n' + genLengthString(38) + '\n' + genLengthString(40) + '\n' + genLengthString(39) + '\n';
        CharacterSource source = CharacterSource.from(str, "<source>");
        PositionMap map = new PositionMap(source, new int[] {0, 39, 80, 120});

        Position pos = map.resolve(50);

        assertSame(50, pos.getIndex());
        assertSame(11, pos.getColumn());
        assertSame(3, pos.getLine());

        com.github.nsc.de.shake.util.CompilerError error =
                new com.github.nsc.de.shake.util.CompilerError(
                        "message", "TestingError", "Some details", pos, pos);

        // System.out.println(error.toString());

        assertEquals("<source>:3:11", error.getMarker().getSource());
        assertEquals("3  012345678901234567890123456789012...+7", error.getMarker().getPreview());
        assertEquals("             ^", error.getMarker().getMarker());

        assertEquals("3  0123456789" + Formatting.INVERT + Formatting.FGColor.RED + "0" +
                Formatting.RESET + "1234567890123456789012...+7", error.getMarker().getColorPreview());

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);

    }

    private static String genLengthString(int length) {
        StringBuilder string = new StringBuilder();
        for(int i = 0; i < length; i++) string.append(i % 10);
        return string.toString();
    }

}

File: src/test/kotlin/com/github/nsc/de/shake/CompilerError.kt
package com.github.shakelang.shake

import com.github.shakelang.shake.util.CompilerError
import com.github.shakelang.shake.util.Formatting
import com.github.shakelang.shake.util.characterinput.charactersource.CharacterSource
import com.github.shakelang.shake.util.characterinput.position.Position
import com.github.shakelang.shake.util.characterinput.position.PositionMap
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class CompilerError {
    @Test
    fun testCompilerError() {
        val source = CharacterSource.from(genLengthString(30), "<source>")
        val map = PositionMap(source, intArrayOf())
        val error = CompilerError(
            "message", "TestingError", "Some details",
            map.resolve(10), map.resolve(10)
        )
        Assertions.assertEquals("<source>:1:11", error.marker.source)
        Assertions.assertEquals("1  012345678901234567890123456789", error.marker.preview)
        Assertions.assertEquals("             ^", error.marker.marker)
        Assertions.assertEquals(
            "1  0123456789" + Formatting.INVERT + Formatting.FGColor.RED + "0" +
                    Formatting.RESET + "1234567890123456789", error.marker.colorPreview
        )

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);
    }

    @Test
    fun testCompilerErrorOverflowAfter() {
        val source = CharacterSource.from(genLengthString(60), "<source>")
        val map = PositionMap(source, intArrayOf())
        val error = CompilerError(
            "message", "TestingError", "Some details",
            Position(map, 10, 11, 1),
            Position(map, 10, 11, 1)
        )

        // System.out.println(error.toString());
        Assertions.assertEquals("<source>:1:11", error.marker.source)
        Assertions.assertEquals("1  0123456789012345678901234567890...+28", error.marker.preview)
        Assertions.assertEquals("             ^", error.marker.marker)
        Assertions.assertEquals(
            "1  0123456789" + Formatting.INVERT + Formatting.FGColor.RED + "0" +
                    Formatting.RESET + "12345678901234567890...+28", error.marker.colorPreview
        )

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);
    }

    @Test
    fun testCompilerErrorOverflowBefore() {
        val source = CharacterSource.from(genLengthString(60), "<source>")
        val map = PositionMap(source, intArrayOf())
        val error = CompilerError(
            "message", "TestingError", "Some details",
            map.resolve(39), map.resolve(39)
        )

        // System.out.println(error.toString());
        Assertions.assertEquals("<source>:1:40", error.marker.source)
        Assertions.assertEquals("1  +18...890123456789012345678901234567890123456789", error.marker.preview)
        Assertions.assertEquals("                              ^", error.marker.marker)
        Assertions.assertEquals(
            "1  +18...890123456789012345678" + Formatting.INVERT + Formatting.FGColor.RED + "9" +
                    Formatting.RESET + "01234567890123456789", error.marker.colorPreview
        )

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);
    }

    @Test
    fun testCompilerErrorOverflow() {
        val source = CharacterSource.from(genLengthString(100), "<source>")
        val map = PositionMap(source, intArrayOf())
        val error = CompilerError(
            "message", "TestingError", "Some details",
            map.resolve(49), map.resolve(49)
        )

        // System.out.println(error.toString());
        Assertions.assertEquals("<source>:1:50", error.marker.source)
        Assertions.assertEquals("1  +28...890123456789012345678901234567890123456789...+29", error.marker.preview)
        Assertions.assertEquals("                              ^", error.marker.marker)
        Assertions.assertEquals(
            "1  +28...890123456789012345678" + Formatting.INVERT + Formatting.FGColor.RED + "9" +
                    Formatting.RESET + "01234567890123456789...+29", error.marker.colorPreview
        )

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);
    }

    @Test
    fun testCompilerErrorLongMark() {
        val source = CharacterSource.from(genLengthString(40), "<source>")
        val map = PositionMap(source, intArrayOf())
        val error = CompilerError(
            "message", "TestingError", "Some details",
            map.resolve(9),
            map.resolve(14)
        )

        // System.out.println(error.toString());
        Assertions.assertEquals("<source>:1:10", error.marker.source)
        Assertions.assertEquals("1  012345678901234567890123456789012345678", error.marker.preview)
        Assertions.assertEquals("            ^^^^^^", error.marker.marker)
        Assertions.assertEquals(
            "1  012345678" + Formatting.INVERT + Formatting.FGColor.RED + "901234" +
                    Formatting.RESET + "567890123456789012345678", error.marker.colorPreview
        )

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);
    }

    @Test
    fun testCompilerErrorMultiLine() {
        val str = """
            
            ${genLengthString(38)}
            ${genLengthString(40)}
            ${genLengthString(39)}
            
            """.trimIndent()
        val source = CharacterSource.from(str, "<source>")
        val map = PositionMap(source, intArrayOf(0, 39, 80, 120))
        val pos = map.resolve(50)
        Assertions.assertSame(50, pos.index)
        Assertions.assertSame(11, pos.column)
        Assertions.assertSame(3, pos.line)
        val error = CompilerError(
            "message", "TestingError", "Some details", pos, pos
        )

        // System.out.println(error.toString());
        Assertions.assertEquals("<source>:3:11", error.marker.source)
        Assertions.assertEquals("3  012345678901234567890123456789012...+7", error.marker.preview)
        Assertions.assertEquals("             ^", error.marker.marker)
        Assertions.assertEquals(
            "3  0123456789" + Formatting.INVERT + Formatting.FGColor.RED + "0" +
                    Formatting.RESET + "1234567890123456789012...+7", error.marker.colorPreview
        )

        // System.out.println(Formatting.FGColor.GREEN + "\u2705 Correct error was thrown" + Formatting.RESET);
    }

    companion object {
        private fun genLengthString(length: Int): String {
            val string = StringBuilder()
            for (i in 0 until length) string.append(i % 10)
            return string.toString()
        }
    }
}